### ISSUE TITLE
keyfile

### ISSUE THREAD

**Issue by NicolasTurpin** [theKeithD/thmj3g-tools#1]
I try to make this works on Touhou Mahjong 4N aving goal to translate it as much as I can !

But I'm limited by the fact that I don't know the differences between 3G files and 4N ones.
the keyfile seems missing... or do I have to create it myself ? What is that keyfile ?

On the ther hand, I think I've understood how I'm supposed to unpack everything, but if I managed to do so, how do I repack everything ?

Thanks a lot !

**Comment by JioPlato** [comment]
Did you ever find it? whats the keyfile

**Comment by JioPlato** [comment]
> I try to make this works on Touhou Mahjong 4N aving goal to translate it as much as I can !
> 
> But I'm limited by the fact that I don't know the differences between 3G files and 4N ones. the keyfile seems missing... or do I have to create it myself ? What is that keyfile ?
> 
> On the ther hand, I think I've understood how I'm supposed to unpack everything, but if I managed to do so, how do I repack everything ?
> 
> Thanks a lot !

if you find the file let me know!

**Comment by remi-sc** [comment]
Finding a keyfile seems fairly easy. For 3G it's in the memory. Even without trying to RE it's possible to get it by dumping the memory after launching, filtering it a bit and brute-force checking in the background until decrypted files return a valid data. it took few-dozen hours of very ad-hoc, non-optimal script to get it in my case. Will try to do the same with 4N.

Since this repo does not include the key, I do not plan to post or link it here to follow the owner's intention.
If you have a different place to discuss the translation then please point me to it.

**Comment by remi-sc** [comment]
yup, 4N key is similarly easy to find - it was even in the same address range as 3G.

So I've got both 3G and 4N keys.
@Sato1999 , @Bobledossier : if you are still interested let me know here.

**Comment by NicolasTurpin** [comment]
@remi-sc OH YEAH !!! I'm very interested in this !
I really want to put my translation in the 4N game, even if it's just for my own achievement haha
You can send me a private message if you want

I'm so excited right now haha
This game means A LOT to me, you can't imagine how much x)

PS: sorry for the close/reopen, it was a missclick

**Comment by remi-sc** [comment]
both keys sent, if anyone else wants them let me know here or send me an email (address on my github profile page.)

**Comment by NicolasTurpin** [comment]
@remi-sc It worked ! I successfully unpacked everything !!
I'll have a look through the files and start testing editing some files and repacking, I hope It will go well !

Damn I'm so happy !! I'll be sure to credit you and theKeithD somewhere ^^

Maybe we can leave this issue open so that the owner can tell us if we can publish the key in here ?

EDIT: Let's goooo ! I figured out how I was supposed to repack everything and now I can see my test modification appearing in the game !
Now I can start implementing everything. I seriously don't know how to thank you and Keith haha
Thanks a lot !

**Comment by NicolasTurpin** [comment]
H‌ey !

I'm nearly done with the translation of Touhou Mahjong 4N, and as you helped me I wanted to credit you somewhere in the game.
My question is : How would you want to be named in there ?
is "remi-sc" enough ? or would you want something else ?

I'm not quite done yet so you have some time to answer me but in any case you'll be in there alongside TheKeithD !

**Comment by remi-sc** [comment]
Yeah, just remi-sc is enough - although I don't really care much about the credit.

I am looking forward to seeing how that translation looks like though!
When I took a cursory look at the decrypted images there was not really that much "text within the images" - which would be easiest to switch by just swapping the contents.

After I saw that there are gigantic kanji image tables I thought the translation would require a lot of RE/binary patching (I presumed the text is glued from extracted image parts)

For example I am extremely curious if / how would yaku names be translated.

**Comment by NicolasTurpin** [comment]
Yeah that's quite a mess in there (actually not that much) but as I had already translated everything beforehand, it helped a bit x3

Some texts (like all the config/records) where inside lua file while others (like character names and powers) where inside CSV files, actually well organized.

The only difficuly here was that characters were far more limited in romaji that actaul japanese characters and lines were cut very strangely so I had to re-write all my translation to fit in the spaces allowed haha.

For the graphic texts, the thing that took me most of my time was finding fonts that matches those used in japanese haha.
I worked in visual effects so cleaning the pictures was quite easy and modifying the picture quite pleasant in some ways ^^
(I'll send you an exemple with this message)

The only road-block I came across was the achievements :

Where Characters and Tables have their own "script name" and "display name" which I could modify,
The Achievements only have one fixed name...
So if I were to change it, it would break the script. And even if I were to modify the entiere script so that every achievement name is in english... the moment you'll play against someone using a japanese game online, the game would break because it wouldn't recognize the Achievement/Title they are using.

So... I'm gonna implement "display name" for the achievements, and even if it seems simple, I'm actually looking for it and hope for the best haha xD



As a side note : I'm french ! I started translating everything to english but I'm not that fluent so... yeah x) Once I've finished everything I'll be more that open to suggestions for a better, more understandable english ;D

See you later !


(PS: Here is an exemple of an icon I had to cleanup : erasing the text while re-drawing the picture and then put the english text with a matching color and font ^^)

(PPS: For the Yaku names : I actually changed them in 2 places > in the Records screen, only in the UI > and in the GFX picture for the Scoring screen... well... for the later I'm still hesitating between 2/3 fonts haha)‌

![Capture 1](https://user-images.githubusercontent.com/62624318/150629980-d5c848d2-13d9-4a3b-bb41-f58c3a993589.png)
![Capture 2](https://user-images.githubusercontent.com/62624318/150629983-a28624b4-574e-4249-ab29-88b95a54a8b3.png)
![Capture 3](https://user-images.githubusercontent.com/62624318/150629984-280b075c-637a-473c-be39-54001288fb1a.png)

PPPS: Few hours later, I have found a nice and pretty solution for the achievements ! there is no more roadblocks until the end ^^
